Fiche utiliser r insee [WIP] (#201)
InseeFrLab/utilitR: 01_R_Insee/Fiche-personnaliser-R.Rmd

# Personnaliser sa configuration

> La personnalisation de R, ou de RStudio doit être réalisée avec précaution, car personnaliser son environnement de travail, c'est prendre le risque de perdre la reproductibilité de ses travaux.

Reviewer: ```suggestion
La personnalisation de `R`, ou de `RStudio` doit être réalisée avec précaution, car personnaliser son environnement de travail, c'est prendre le risque de rendre moins reproductible ses travaux.

```

Il convient donc de bien comprendre ce que la personnalisation peut apporter, et ce qu'elle ne doit surtout pas faire.

::: {.recommandation data-latex=""}

> Ne personnalisez pas R, sauf dans des cas très précis :

Reviewer: ```suggestion
Ne personnalisez pas `R`, sauf dans des cas très précis :

```

Author: A titre très personnel, je trouve que cette typographie de R (dans tout le bouquin) dans le style informatique rend la lecture plus compliquée. On sait que R est un logiciel statistique, pas besoin de le mettre en valeur (idem pour RStudio). 
Mais je comprends également votre position, et cette fiche sera harmonisée avec les autres 😸

* pour masquer des informations qui **ont vocation à être masquées**, comme des clés d'API dans le fichier `.Renviron` ;

> * pour modifier des paramètres qui n'ont pas vocation à être bien configurés dans R à l'Insee, comme le proxy ou des chemins vers executables ;

Reviewer: ```suggestion
* pour modifier des paramètres qui n'ont pas vocation à être bien configurés dans `R` à l'Insee, comme le proxy ou des chemins vers executables ;

```

> * pour changer des éléments qui n'ont aucun impact sur les calculs, mais peuvent vous faciliter la vie, comme l'augmentation de la taille de l'historique, le repos, le prompt, etc.

Reviewer: > pour changer des éléments qui n'ont aucun impact sur les calculs, mais peuvent vous faciliter la vie, comme l'augmentation de la taille de l'historique, le repos, le prompt, etc.

Qu'est-ce que tu entends par repos ? L'adresse des dépôts de package type miroir du CRAN ?

Author: Oui, exactement, l'adresse des dépots de packages.

D'autres paramètres, notamment ceux de RStudio, sont aisément modifiables, mais parfois non automatisable (comme le changement d'apparence).
:::

## Comprendre le lancement de R ?

> R a été conçu, dès le début, pour permettre une très grande flexibilité d'utilisation. Cette flexibilité engendre toutefois une complexité dans ses procédures d'initialisation.  La documentation sur le lancement de R montre d'ailleurs toute la complexité de la chose :

Reviewer: ```suggestion
`R` a été conçu, dès le début, pour permettre une très grande flexibilité d'utilisation. Cette flexibilité engendre toutefois une complexité dans ses procédures d'initialisation.  [La documentation sur le lancement de `R`](https://cran.r-project.org/web/packages/startup/vignettes/startup-intro.html) montre d'ailleurs toute la complexité de la chose :

```

```{r, eval = FALSE}
help("Startup")
```

Néanmoins, et pour simplifier grandement, la procédure se déroule en trois temps :

1. R cherche des fichiers de type *environment*. Ces fichiers permettent notamment d'indiquer des chemins spécifiques où se trouvent d'autres fichiers, ou certaines caractéristiques comme le proxy.
2. R cherche des fichiers de type *profile*. Ces fichiers contiennent du code R qui se lance automatiquement. Ils permettent par exemple d'indiquer des packages supplémentaires à charger, de personnaliser le message d'accueil de R, ou de sauvegarder automatiquement les informations de la session lorsqu'on la quitte.

> 3. R prépare la session, avec les informations récupérées précédemment, mais également d'autres, comme par exemple les données qu'il aurait pu conserver dans un .Rdata, puis il charge les packages de base, avant par exemple de charger l'historique des commandes (s'il existe). C'est grace à ce dernier temps que R recharge les données et l'historique après s'être interrompu inopunément.

Reviewer: ```suggestion
3. R prépare la session, avec les informations récupérées précédemment, mais également d'autres, comme par exemple les données qu'il aurait pu conserver dans un `.Rdata`, puis il charge les packages de base, avant par exemple de charger l'historique des commandes (s'il existe). C'est grâce à ce dernier temps que R recharge les données et l'historique après s'être interrompu inopunément.

```

> Sans complexifier outre mesure, sachez que les deux premiers temps se décomposent eux-mêmes en deux sous-temps : il est possible de définir des fichiers *environment* ou *profile* plus génériques que d'autres (c'est le `Renviron.site` ou `Rprofile.site`).

Reviewer: Il faut expliquer le système en poupée russe des .Renviron et .Rprofile: dans la racine de R > dans un projet Rstudio

* il définit sa clé d'API dans une variable MON_API_INSEE . Ultérieurement le code pourra faire appel à cette variable via `Sys.getenv("MON_API_INSEE")`. Ce faisant la clé confidentielle n'est pas directement visible dans le code, et ce dernier peut donc être librement partagé sans risque de sécurité. Comme le fichier `.Renviron` peut contenir des informations sensibles, il ne doit pas être partagé (notamment via Git !) ;
* il modifie les paramètres du proxy pour permettre à R de se connecter à Internet&nbsp;;
* il modifie des variables internes à R (`_R_CHECK_LENGTH_1_CONDITION_`  et `_R_CHECK_LENGTH_1_LOGIC2_`)&nbsp;;
* il **contient une ligne vide finale** : cette ligne vide est indispensable, ne l'oubliez jamais !

Pour modifier le fichier `.Renviron`, le plus simple est sans doute de procéder ainsi :

1. Tapez la commande `usethis::edit_r_environ()`, le fichier `.Renviron` de votre profil utilisateur s'ouvrira. Si vous travaillez dans un *RProject*, l'option `scope="project"` permet de définir des options spécifiques à votre projet.
2. Modifiez le fichier, sans oublier la ligne vide en fin de document. Sauvegardez le.

> 3. Relancez R, via *Session/Restart* ou *Ctrl+Shift+F10*

Reviewer: >  3. Relancez R, via *Session/Restart* ou *Ctrl+Shift+F10* 

@oliviermeslin : tu pourras formatter les touches comme tu l'as fait à d'autres endroits?

::: {.remarque data-latex=""}
La modification de variables internes à R est à utiliser avec précaution, car elle est susceptible de générer de la non-reproductibilité. Par exemple, R a un comportement assez étrange dans le cas suivant : 
```{r}
vecteurboolean <- c(TRUE, FALSE, NULL)
if (vecteurboolean) {
  cat("Je suis vrai ! \n")
} else {
  cat("Je suis faux ! \n")
}
```

> Lorsqu'il teste un vecteur de booléen, R se contente de tester la première valeur, et met un simple avertissement. Toutefois, avec la modification de `.Renviron` de type `_R_CHECK_LENGTH_1_CONDITION_ = "TRUE"`, il est possible de modifier ce comportement et de générer une erreur (qui stoppera le code), si on fournit à R un vecteur au lieu de fournir un simple scalaire. Le code précédent... n'imprimera donc rien du tout et s'arretera avec une erreur ! Vous comprenez mieux les raisons pour lesquels modifier l'environnement peut conduire à des résultats non reproductibles ! 

Reviewer: ```suggestion
Lorsqu'il teste un vecteur de booléen, R se contente de tester la première valeur, et met un simple avertissement. Toutefois, avec la modification de `.Renviron` de type `_R_CHECK_LENGTH_1_CONDITION_ = "TRUE"`, il est possible de modifier ce comportement et de générer une erreur (qui stoppera le code), si on fournit à R un vecteur au lieu de fournir un simple scalaire. Le code précédent... n'imprimera donc rien du tout et s’arrêtera avec une erreur chez vous ! Chez un autre utilisateur, il n'y aura pas d'erreur. Vous comprenez mieux les raisons pour lesquels modifier l'environnement peut conduire à des résultats non reproductibles ! 

```

:::

Il est toutefois possible de modifier les paramètres de son environnement ou en générer d'autres, une fois R chargé, via la commande `Sys.setenv()`. par exemple `Sys.setenv("MON_API_INSEE"="jamais")`. De même la fonction `Sys.getenv()` permet de lister certaines valeurs présentes dans l'environnement actuel. Toutefois, toutes les procédures d'initialisation de R se seront déjà déroulées alors avec les anciennes valeurs, définies par défault. Cette subtilité peut être importante !

::: {.remarque data-latex=""}

> La navigation sur Internet depuis un poste de l'Insee est contrôlée par un *proxy* qu'il faut paramétrer pour accéder à Internet.

Reviewer: ```suggestion
La navigation sur Internet depuis un poste de l'Insee, comme dans de nombreuses institutions, est contrôlée par un *proxy* (intermédiaire entre le *web* et un ordinateur) qu'il faut paramétrer pour accéder à Internet.

```

Pour se faire, il est recommandé de récupérer l'adresse du proxy interne en réalisant la commande suivante :

```{r,eval=FALSE}
curl::ie_get_proxy_for_url()
```

Puis, vous pourrez mettre à jour votre `.Renviron` via les variables d'environnement `http_proxy`et `https_proxy`.
:::

### .Rprofile

Le fichier `.RProfile` contient du code R qui sera lancé par R après les fichiers `.Renviron`. Typiquement, on peut utiliser ce fichier pour :

* modifier le miroir par défault du CRAN
* modifier le message d'accueil
* personnaliser l'affichage de R

> Un fichier `.Rprofile` pourrait typiquement ressembler à quelque chose comme :

Reviewer: Tu peux faire un lien avec le .Rprofile du livre: impression d'un message pour les contributeurs

(il y a des fonctions, c'est pas très propre, mais à terme mais on va les mettre dans un package spécifique #218)

```{r, eval=FALSE}
options(repos = c(CRAN = "https://cran.rstudio.org"))
options(prompt="R> ", digits=4)
options(continue= "+++ ")
if (interactive()) {
options(width = 120)
}

.First <- function(){
cat("\n Allez, hop hop hop, au travail !\n\n")
}
.Last <- function(){
cat("\n Je m'en vais comme un prince...\n\n")
}

```
Ce fichier permet de :

* définir un miroir par défault pour le CRAN ;
* modifier l'affichage de l'invite de commande (le ">" de début de ligne devient "R> ") ;
* modifier l'affichage des nombres décimaux  ;
* modifier l'affichage de la console lorsqu'une partie du code reste à écrire (le "+ " en début de ligne devient "+++ ") ;
* dans le cadre d'un appel à la console, la largeur des impressions est fixée à 120 ;
* les fonctions `.First` et `.Last` permettent d'effectuer des opérations spécifiques au lancement et à la fermeture de la session R (ici imprimer du texte) ;
* Attention, comme précédemment, **un fichier `.Rprofile` doit se terminer par une ligne vierge.** 

> Pour voir certaines options aisément modifiables, la fonction `options()` affiche les valeurs actuellement chargées, et il est possible de les modifier à la volée via `options(nom_de_l_option=valeur_de_l_option)`.

Reviewer: Il me semble qu'`options` ne survit pas à une session `R`. Ca vaut le coup de préciser la nuance

* définir des alias de fonctions ;
* modifier le comportement par défault, comme par exemple `options(stringAsFactors=FALSE)` ;

En effet, toute personne sans ce fichier de configuration ne sera pas en mesure de reproduire vos travaux.

Si vous réalisez des opérations spécifiques dans votre `.Rprofile`, il est conseillé de :

* travailler dans le `.Rprofile` d'un projet, et non pas le `.Rprofile` général de votre profil d'utilisateur.
* partager le fichier `.Rprofile` du projet, après avoir vérifié qu'il ne contenait aucune information sensible.

:::

## Quelques bonnes pratiques

La personnalisation de R est une opération sensible qui ne doit pas interférer avec les obligations de reproductibilité des travaux statistiques. Il convient de réserver cette personnalisation à des modifications sans impact sur les calculs età des informations ayant vocation à ne pas être diffusées : 

* pour des raisons de sécurité : pour séparer les informations sensibles (comme les identifiants d'API ou les paramètres de proxy) du code librement partagé ;
* pour des raisons d'infrastructure : pour modifier les chemins vers des executables, en raison des droits limités sur les machines.

> Les autres opérations modifiant le comportement de base (comme `stringAsFactors=FALSE`) doivent être partagées et accessibles.

Reviewer: Préciser : plus besoin de faire ça dans R 4.0 car c'est devenu le comportement par défaut

1. Les fichiers présents dans un `Rproject` sont prioritaires.
2. A défault, les fichiers présents dans le _Home_ (sous Windows, ce sera typiquement _Mes documents_).
3. A défault, les fichiers présents dans le répertoire d'installation de `R`.

Attention, R ne travaille qu'avec un unique fichier *environment* et un unique fichier *profile*. Créer un nouveau fichier *environment* dans un projet désactivera donc l'éventuel fichier présent dans le _Home_. 

### .Renviron

Le fichier d'environnement `.Renviron` sert à définir des... variables d'environnement ! Ces variables sont utiles à l'interaction entre R et d'autres processus, ou uniquemnent à R. Ces variables d'environnement peuvent également modifier la manière dont R réalise certains calculs. Encore une fois, il ne faut pas faire n'importe quoi !

Par exemple, un `.Renviron` fictif pourrait ressembler à : 

```
MON_API_INSEE=MaCleSecreteVraimentTresSecrete
http_proxy=le_proxy
https_proxy=le_proxy
_R_CHECK_LENGTH_1_LOGIC2_=verbose
_R_CHECK_LENGTH_1_CONDITION_=true

```

> Comme vous pouvez le constater, un tel fichier n'est pas constitué de code R, mais de paires **nom=valeur**. Chaque paire définit une variable d'environnement qui sera exploitée par R pour interagir avec d'autres logiciels, ou directement pour lui. Ce fichier fictif réalise les opérations suivantes :

Author: résolu manuellement, j'avais oublié de cliquer sur cette suggestion de commit.

> * il définit sa clé d'API dans une variable MON_API_INSEE . Ultérieurement le code pourra faire appel à cette variable via `Sys.getenv("MON_API_INSEE")`. Ce faisant la clé confidentielle n'est pas directement visible dans le code, et ce dernier peut donc être librement partagé sans risque de sécurité. Comme le fichier `.Renviron` peut contenir des informations sensibles, il ne doit pas être partagé (notamment via Git !) ;

Author: idem, résolution manuelle.

> * il modifie les paramètres du proxy pour permettre à R de se connecter à Internet&nbsp;;
> * il modifie des variables internes à R (`_R_CHECK_LENGTH_1_CONDITION_`  et `_R_CHECK_LENGTH_1_LOGIC2_`)&nbsp;;
> * il **contient une ligne vide finale** : cette ligne vide est indispensable, ne l'oubliez jamais !

Reviewer 2: ```suggestion
* il modifie les paramètres du _proxy_ pour permettre à `R` de se connecter à Internet ;
* il modifie des variables internes à `R` (telles que `_R_CHECK_LENGTH_1_CONDITION_`  et `_R_CHECK_LENGTH_1_LOGIC2_`) ;
* il **contient une ligne vide finale** : cette ligne vide est indispensable, ne l'oubliez jamais !

```

> Pour modifier le fichier `.Renviron`, le plus simple est sans doute de procéder ainsi :
> 
> 1. Tapez la commande `usethis::edit_r_environ()`, le fichier `.Renviron` de votre profil utilisateur s'ouvrira. Si vous travaillez dans un *RProject*, l'option `scope="project"` permet de définir des options spécifiques à votre projet.
> 2. Modifiez le fichier, sans oublier la ligne vide en fin de document. Sauvegardez le.
> 3. Relancez R, via *Session/Restart* ou *Ctrl+Shift+F10*

Reviewer 2: ```suggestion
La méthode la plus simple pour modifier le fichier `.Renviron` est la suivante :

1. Ouvrir le fichier `.Renviron`. La commande `usethis::edit_r_environ()` permet d'ouvrir le fichier `.Renviron` de votre profil utilisateur (c'est le fichier d'environnement utilisé lorsqu'il n'y a pas de fichier d'environnement à votre projet). Si vous travaillez dans un `RProject`, la commande `usethis::edit_r_environ(scope="project")` permet d'ouvrir le le fichier `.Renviron` spécifique à votre projet.
2. Modifier le fichier, sans oublier la ligne vide en fin de document. 
3. Sauvegarder le fichier.
3. Relancez `R`, via `Session > Restart R` ou <kbd>Ctrl</kbd> + <kbd>⇧ Shift</kbd> + <kbd>F10</kbd>.
```

::: {.remarque data-latex=""}

> La modification de variables internes à R est à utiliser avec précaution, car elle est susceptible de générer de la non-reproductibilité. Par exemple, R a un comportement assez étrange dans le cas suivant : 
> ```{r}
> vecteurboolean <- c(TRUE, FALSE, NULL)
> if (vecteurboolean) {
>   cat("Je suis vrai ! \n")
> } else {
>   cat("Je suis faux ! \n")
> }
> ```

Reviewer 2: ```suggestion
Il faut être prudent lorsqu'on modifie des variables internes à `R`, car ces modifications peuvent aboutir à ce que les résultats des scripts ne soient pas reproductibles. Par exemple, `R` a un comportement assez étrange dans le cas suivant : 
```{r}
vecteurboolean <- c(TRUE, FALSE, NULL)
if (vecteurboolean) {
  cat("Je suis vrai ! \n")
} else {
  cat("Je suis faux ! \n")
}
```

```

> Lorsqu'il teste un vecteur de booléen, R se contente de tester la première valeur, et met un simple avertissement. Toutefois, avec la modification de `.Renviron` de type `_R_CHECK_LENGTH_1_CONDITION_ = "TRUE"`, il est possible de modifier ce comportement et de générer une erreur (qui stoppera le code), si on fournit à R un vecteur au lieu de fournir un simple scalaire. Le code précédent... n'imprimera donc rien du tout et s’arrêtera avec une erreur chez vous ! Chez un autre utilisateur, il n'y aura pas d'erreur. Vous comprenez mieux les raisons pour lesquels modifier l'environnement peut conduire à des résultats non reproductibles ! 
> 

Reviewer 2: ```suggestion
Lorsqu'une condition porte sur un vecteur de booléens (`TRUE` ou `FALSE`), `R` utilise uniquement le premier élément du vecteur pour vérifier si condition est remplie, et renvoie un simple avertissement à l'utilisateur. Toutefois, il est possible de modifier ce comportement par défaut en indiquant `_R_CHECK_LENGTH_1_CONDITION_ = "TRUE"` dans le fichier `.Renviron` : `R` renverra alors une erreur et stoppera le code si `R` doit vérifier une condition sur un vecteur au lieu d'un simple scalaire. Dans ce cas, le code précédent ne renverra rien du tout et s’arrêtera avec une erreur. En revanche, chez un autre utilisateur dont le fichier `.Renviron` ne contient pas ce réglage, il n'y aura pas d'erreur. Le même code aboutit donc à des résultats différents en raison de différences dans les variables d'environnement.
```

:::

> Il est toutefois possible de modifier les paramètres de son environnement ou en générer d'autres, une fois R chargé, via la commande `Sys.setenv()`. par exemple `Sys.setenv("MON_API_INSEE"="jamais")`. De même la fonction `Sys.getenv()` permet de lister certaines valeurs présentes dans l'environnement actuel. Toutefois, toutes les procédures d'initialisation de R se seront déjà déroulées alors avec les anciennes valeurs, définies par défault. Cette subtilité peut être importante !
> 
> ::: {.remarque data-latex=""}
> La navigation sur Internet depuis un poste de l'Insee, comme dans de nombreuses institutions, est contrôlée par un *proxy* (intermédiaire entre le *web* et un ordinateur) qu'il faut paramétrer pour accéder à Internet.
> 
> Pour ce faire, il est recommandé de récupérer l'adresse du proxy interne en réalisant la commande suivante :
> 
> ```{r,eval=FALSE}
> curl::ie_get_proxy_for_url()
> ```
> 
> Puis, vous pourrez mettre à jour votre `.Renviron` via les variables d'environnement `http_proxy`et `https_proxy`.
> :::

Reviewer 2: ```suggestion
Il est possible de modifier les paramètres de son environnement ou d'en générer d'autres après l'ouverture de la session `R`, avec la commande `Sys.setenv()`. Vous pouvez par exemple exécuter la commande `Sys.setenv("MON_API_INSEE"="jamais")`. De même la fonction `Sys.getenv()` permet de lister certaines valeurs présentes dans l'environnement actuel. Toutefois, toutes les procédures d'initialisation de R se seront déjà déroulées alors avec les anciennes valeurs, définies par défaut. Cette subtilité peut être importante !

::: {.remarque data-latex=""}
Comme dans de nombreuses institutions, la navigation sur Internet depuis un poste de l'Insee est contrôlée par un *proxy* (intermédiaire entre le *web* et un ordinateur). Il est indispensable de paramétrer l'adresse du _proxy_ pour que `R` puisse accéder à Internet (par exemple pour télécharger un _package_). Pour ce faire, vous pouvez récupérer l'adresse du _proxy_ interne en exécutant la commande suivante :
```{r, eval=FALSE}
curl::ie_get_proxy_for_url()
```
Puis vous pouvez ajouter l'adresse du _proxy_ dans votre fichier `.Renviron`. Il faut ajouter deux lignes dans ce fichier :
```{r, eval = FALSE}
http_proxy=adresse_du_proxy
https_proxy=adresse_du_proxy
```
:::

```

> ### .Rprofile

Reviewer 2: ```suggestion
### Le fichier `.Rprofile`

```

> Le fichier `.RProfile` contient du code R qui sera lancé par R après les fichiers `.Renviron`. Typiquement, on peut utiliser ce fichier pour :
> 
> * modifier le miroir par défault du CRAN
> * modifier le message d'accueil
> * personnaliser l'affichage de R
> 
> Un fichier `.Rprofile` pourrait typiquement ressembler à quelque chose comme :

Reviewer 2: ```suggestion
Le fichier `.RProfile` contient du code `R` qui sera exécuté par `R` après le chargement des fichiers `.Renviron`. Voici trois utilisations standards du fichier `.Rprofile` :

* modifier le miroir par défaut du CRAN (pour télécharger des _packages_) ;
* modifier le message d'accueil de la session ;
* personnaliser l'affichage de `R`.

Un fichier `.Rprofile` peut par exemple ressembler à ceci :

```

```
options(repos = c(CRAN = "https://cran.rstudio.org"))
options(prompt="R> ", digits=4)
options(continue= "+++ ")
if (interactive()) {
options(width = 120)
}

.First <- function(){
cat("\n Allez, hop hop hop, au travail !\n\n")
}
.Last <- function(){
cat("\n Je m'en vais comme un prince...\n\n")
}

```

> Ce fichier permet de :
> 
> * définir un miroir par défault pour le CRAN ;
> * modifier l'affichage de l'invite de commande (le ">" de début de ligne devient "R> ") ;
> * modifier l'affichage des nombres décimaux  ;
> * modifier l'affichage de la console lorsqu'une partie du code reste à écrire (le "+ " en début de ligne devient "+++ ") ;
> * dans le cadre d'un appel à la console, la largeur des impressions est fixée à 120 ;
> * les fonctions `.First` et `.Last` permettent d'effectuer des opérations spécifiques au lancement et à la fermeture de la session R (ici imprimer du texte) ;
> * Attention, comme précédemment, **un fichier `.Rprofile` doit se terminer par une ligne vierge.** 
> 
> Pour voir certaines options aisément modifiables, la fonction `options()` affiche les valeurs actuellement chargées, et il est possible de les modifier à la volée via `options(nom_de_l_option=valeur_de_l_option)`. Attention, la fonction `options()` n'est opérationnelle que sur la session de `R` actuellement active ! Si `R` redemarre, toutes les options modifiées à la volée ne sont plus prises en compte.
> 
> Un exemple plus réaliste de `.Rprofile` peut être trouvé [ici](https://github.com/InseeFrLab/utilitR/blob/master/.Rprofile) : c'est le fichier qui permet notamment de réaliser la documentation d'UtilitR. Notez notamment le message à l'ouveture du projet ! Avoir des fonctions n'est pas forcément interdit, tant que ces fonctions sont également partagées, comme c'est le cas dans ce projet !

Reviewer 2: ```suggestion
Ce fichier `.Rprofile` permet de :

* définir un miroir par défaut pour le CRAN ;
* modifier l'affichage de l'invite de commande (le ">" de début de ligne devient "R> ") ;
* modifier l'affichage par défaut des nombres décimaux  ;
* modifier l'affichage de la console lorsqu'une partie du code reste à écrire (le "+ " en début de ligne devient "+++ ") ;
* dans le cadre d'un appel à la console, la largeur des impressions est fixée à 120 ;
* les fonctions `.First` et `.Last` permettent d'effectuer des opérations spécifiques au lancement et à la fermeture de la session R (imprimer du texte dans le cas présent) ;
* Attention, comme précédemment, **un fichier `.Rprofile` doit se terminer par une ligne vierge**.

La fonction `options()` affiche les valeurs actuellement chargées pour les options. Il est possible de les modifier à la volée avec la commande `options(nom_de_l_option=valeur_de_l_option)`. Attention toutefois, modifier une option avec la fonction `options()` n'est valable que pour la session de `R` actuellement active ! Si `R` redémarre, toutes les options modifiées à la volée ne sont plus prises en compte. Si vous voulez modifier durablement la valeur d'une option, il est préférable de modifier le fichier `.Rprofile`.

Un exemple plus réaliste de `.Rprofile` peut être trouvé [ici](https://github.com/InseeFrLab/utilitR/blob/master/.Rprofile) : c'est le fichier qui permet notamment de réaliser la documentation d'`utilitR`. Notez notamment le message à l'ouverture du projet ! Avoir des fonctions n'est pas forcément interdit, tant que ces fonctions sont également partagées, comme c'est le cas dans ce projet !

```

> ::: {.conseil data-latex=""}
> 
> Il est ainsi fortement déconseillé de réaliser les choses suivantes dans le `.Rprofile`:
> 
> * charger des packages ;
> * définir des alias de fonctions ;
> * modifier le comportement par défault, comme par exemple `options(stringAsFactors=FALSE)` ;
> 
> En effet, toute personne sans ce fichier de configuration ne sera pas en mesure de reproduire vos travaux.
> 
> Si vous réalisez des opérations spécifiques dans votre `.Rprofile`, il est conseillé de :
> 
> * travailler dans le `.Rprofile` d'un projet, et non pas le `.Rprofile` général de votre profil d'utilisateur.
> * partager le fichier `.Rprofile` du projet, après avoir vérifié qu'il ne contenait aucune information sensible.
> 
> :::
> 

Reviewer 2: ```suggestion
::: {.conseil data-latex=""}

Si vous définissez des réglages spécifiques dans votre `.Rprofile`, il est fortement conseillé de :

* travailler dans le `.Rprofile` d'un projet, et non pas le `.Rprofile` général de votre profil d'utilisateur ;
* partager le fichier `.Rprofile` du projet, après avoir vérifié qu'il ne contenait aucune information sensible.

**Il est fortement déconseillé de réaliser les choses suivantes dans le fichier `.Rprofile`**:

* charger des _packages_ ;
* définir des alias de fonctions ;
* modifier le comportement par défaut de `R`, comme par exemple `options(stringAsFactors=FALSE)`.

En effet, toute personne qui ne dispose pas de ce fichier de configuration ne sera pas en mesure de reproduire vos travaux.
:::
```

> ## Quelques bonnes pratiques

Reviewer 2: Deux points:

- Je me demande si la partie bonnes pratiques ne devrait pas être en premier, juste après la recommandation.
- Je me demande si on ne devrait pas donner une règle précise sur le partage des fichiers de config, par exemple: ne __jamais__ partager le `.Renviron`, et partager le `.Rprofile` uniquement lorsque c'est celui d'un projet (le `.Rprofile` du profil utilisateur c'est pas important). Ca pourrait même être dans les recommandations. @linogaliana : un avis?

Author: Mouais, mais la structure des fiches est telle que normalement, on conclut par cette partie...

Reviewer 2: OK

> La personnalisation de R est une opération sensible qui ne doit pas interférer avec les obligations de reproductibilité des travaux statistiques. Il convient de réserver cette personnalisation à des modifications sans impact sur les calculs et à des informations ayant vocation à ne pas être diffusées : 
> 
> * pour des raisons de sécurité : pour séparer les informations sensibles (comme les identifiants d'API ou les paramètres de proxy) du code librement partagé ;
> * pour des raisons d'infrastructure : pour modifier les chemins vers des executables, en raison des droits limités sur les machines.
> 
> 
> Les autres opérations modifiant le comportement de base (comme `stringAsFactors=FALSE`) doivent être partagées et accessibles.

Reviewer 2: ```suggestion
**La personnalisation de `R` est une opération sensible qui ne doit pas interférer avec l'exigence de reproductibilité des travaux statistiques.** Cette personnalisation doit donc porter uniquement sur des modifications sans impact sur les calculs et sur des informations ayant vocation à ne pas être diffusées. Ces informations à ne pas diffuser peuvent être de deux types :

* **informations sensibles** (comme les identifiants d'API ou les paramètres de proxy). Il est important que ces informations ne figurent pas dans les programmes de façon à pouvoir partager librement les codes ;
* **informations d'infrastructure** (comme les chemins vers des exécutables). Ces informations n'ont pas vocation à être partagées, car elles varient d'un poste d'un agent à l'autre.

Les autres opérations modifiant le comportement de base (comme `stringAsFactors=FALSE`) doivent être partagées et accessibles, _via_ le fichier `.Rprofile`.
```

> ## Pour en savoir plus
> 
> L'immense majorité de la documentation sur cette exploitation de R avancée est en langue anglaise : 
> 
> * [Using .Rprofile and .Renviron, by Nicholas Tierney](https://github.com/numbats/resources/blob/b2fdc170277ecebea9a6358d7d222b55a206d70e/rprofile-renviron.Rmd) ;
> * [Rstats What they forgot to teach you - chapter 7, by Jennifer Bryan \& Jim Hester](https://rstats.wtf/r-startup.html) ;
> * [Managing R with .Rprofile, .Renviron, Rprofile.site, Renviron.site, rsession.conf, and repos.conf by Alex Gold](https://support.rstudio.com/hc/en-us/articles/360047157094-Managing-R-with-Rprofile-Renviron-Rprofile-site-Renviron-site-rsession-conf-and-repos-conf) ;
> * [Efficient R Programing - chapter 2.5, by Colin Gillespie \& Robin Lovelace](https://csgillespie.github.io/efficientR/set-up.html#r-startup) ;
> * [R for Enterprise: Understanding R’s Startup, by Sean Lopp (RStudio)](https://rviews.rstudio.com/2017/04/19/r-for-enterprise-understanding-r-s-startup/)

Reviewer 2: ```suggestion
## Pour en savoir plus

L'immense majorité de la documentation sur l'utilisation des fichiers de configuration de `R` est en langue anglaise : 

* [Using `.Rprofile` and .`Renviron`](https://github.com/numbats/resources/blob/b2fdc170277ecebea9a6358d7d222b55a206d70e/rprofile-renviron.Rmd) par Nicholas Tierney ;
* [R Startup](https://rstats.wtf/r-startup.html), chapitre 7 de _Rstats - What they forgot to teach you_, par Jennifer Bryan er Jim Hester ;
* [Managing `R` with `.Rprofile`, `.Renviron`, `Rprofile.site`, `Renviron.site`, `rsession.conf`, and `repos.conf`](https://support.rstudio.com/hc/en-us/articles/360047157094-Managing-R-with-Rprofile-Renviron-Rprofile-site-Renviron-site-rsession-conf-and-repos-conf) par Alex Gold ;
* [Efficient `R` Programing - chapter 2.5](https://csgillespie.github.io/efficientR/set-up.html#r-startup) par Colin Gillespie \& Robin Lovelace ;
* [`R` for Enterprise: Understanding R’s Startup](https://rviews.rstudio.com/2017/04/19/r-for-enterprise-understanding-r-s-startup/) par Sean Lopp (RStudio).
```
